### PR TITLE
Optionally return all parse errors.

### PR DESCRIPTION
**Pull request by ddunkin**
There could be multiple errors in a service definition. Ideally the language server would report all errors, not just the first one. Syntactic errors may be hard to parse through, but semantic errors should be relatively easy to collect.

This PR makes validation while parsing optional internally, allowing an invalid service to be constructed, but returning all semantic errors. A semantically invalid service is still useful in cases like a language server handling a completion request.

### CODE REVIEW

File: src/Facility.Definition/Fsd/FsdParser.cs [hunk 1]
```diff
@@ -15,8 +15,24 @@ public sealed class FsdParser
...
+				return service;
+			else
+				throw errors.First().CreateException();
+		}
+
+		/// <summary>
+		/// Parses an FSD file into a service definition.
+		/// </summary>
+		/// <returns>true if the parse succeeds.</returns>
+		public bool TryParseDefinition(NamedText source, out ServiceInfo service, out List<ServiceDefinitionError> errors)
```
Comment (ejball) [inline src/Facility.Definition/Fsd/FsdParser.cs]: `IReadOnlyList<ServiceDefinitionError>`

File: src/Facility.Definition/Fsd/FsdParser.cs [hunk 2]
```diff
@@ -49,8 +65,9 @@ public ServiceInfo ParseDefinition(NamedText source)
 
 							var position = new NamedTextPosition(source.Name, headingLineNumber, 1);
 							if (remarksSections.ContainsKey(name))
-								throw new ServiceDefinitionException("Duplicate remarks heading: " + name, position);
-							remarksSections.Add(name, new FsdRemarksSection(name, lines, position));
+								errors.Add(new ServiceDefinitionError("Duplicate remarks heading: " + name, position));
+							else
+								remarksSections[name] = new FsdRemarksSection(name, lines, position);
```
Comment (ejball) [inline src/Facility.Definition/Fsd/FsdParser.cs]: Why change from `Add`?

File: src/Facility.Definition/Fsd/FsdParser.cs [hunk 3]
```diff
@@ -86,21 +110,14 @@ public ServiceInfo ParseDefinition(NamedText source)
...
-			foreach (var remarksSection in remarksSections.Values)
-			{
-				string sectionName = remarksSection.Name;
-				if (service.Name != sectionName && service.FindMember(sectionName) == null)
-					throw new ServiceDefinitionException($"Unused remarks heading: {sectionName}", remarksSection.Position);
+					exception));
 			}
 
-			return service;
+			return errors.Count == 0;
```
Comment (ejball) [inline src/Facility.Definition/Fsd/FsdParser.cs]: Should we `null` the `service` if there are any errors?

File: src/Facility.Definition/ServiceDefinitionError.cs [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+		/// <summary>
+		/// The position where the error took place, if any.
+		/// </summary>
+		public NamedTextPosition Position { get; }
+
+		/// <summary>
+		/// The exception that caused the error, if any.
+		/// </summary>
+		public Exception InnerException { get; }
```
Comment (ejball) [inline src/Facility.Definition/ServiceDefinitionError.cs]: Just `Exception` seems fine to me.

File: src/Facility.Definition/ServiceDefinitionError.cs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+			Error = error;
+			Position = position;
+			InnerException = innerException;
+		}
+
+		/// <summary>
+		/// The error message.
+		/// </summary>
+		public string Error { get; }
```
Comment (ejball) [inline src/Facility.Definition/ServiceDefinitionError.cs]: `Message`?

File: src/Facility.Definition/ServiceDefinitionError.cs [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+		/// <summary>
+		/// The position where the error took place, if any.
```
Comment (ejball) [inline src/Facility.Definition/ServiceDefinitionError.cs]: Is the position ever unspecified?

File: build.cake [hunk 7]
```diff
@@ -44,7 +44,7 @@ Task("Clean")
 	});
 
 Task("Build")
-	.IsDependentOn("Clean")
+//	.IsDependentOn("Clean")
```
Comment (ejball) [inline build.cake]: No objection to dropping this.

File: build.cake [hunk 8]
```diff
@@ -60,7 +60,8 @@ Task("VerifyCodeGen")
 	.Does(() => CodeGen(verify: true));
 
 Task("Test")
-	.IsDependentOn("VerifyCodeGen")
+	.IsDependentOn("Build")
+//	.IsDependentOn("VerifyCodeGen")
```
Comment (ejball) [inline build.cake]: Fair, as long as it gets run during automated builds. Add to `SourceIndex`?

Comment (ddunkin) [reply build.cake]: Accidental commit.

File: src/Facility.Definition/Fsd/FsdParser.cs (hunk 2 above)
Comment (ddunkin) [reply src/Facility.Definition/Fsd/FsdParser.cs]: I had made it last-one-wins, then changed it back (by adding the else) and forgot to change this.

File: src/Facility.Definition/ServiceDefinitionError.cs (hunk 6 above)
Comment (ddunkin) [reply src/Facility.Definition/ServiceDefinitionError.cs]: There is one place it isn't, but that can be changed.

File: src/Facility.Definition/Fsd/FsdParser.cs (hunk 3 above)
Comment (ddunkin) [reply src/Facility.Definition/Fsd/FsdParser.cs]: The idea is that the service can be returned, even if semantically invalid, so that it can be used for completion by the language service.

File: SolutionInfo.cs [hunk 9]
```diff
@@ -1,6 +1,6 @@
 ﻿using System.Reflection;
 
-[assembly: AssemblyVersion("1.2.0.0")]
+[assembly: AssemblyVersion("1.3.0.0")]
```
Comment (ejball) [inline SolutionInfo.cs]: This will need to be bumped.

File: src/Facility.Definition/ServiceDefinitionUtility.cs [hunk 10]
```diff
@@ -95,7 +95,8 @@ internal static IEnumerable<ServiceDefinitionError> ValidateName(string name, Na
 
 		internal static IEnumerable<ServiceDefinitionError> ValidateTypeName(string name, NamedTextPosition position)
 		{
-			if (ServiceTypeInfo.TryParse(name, x => s_validNameRegex.IsMatch(x) ? new ServiceDtoInfo(x) : null, position, out ServiceDefinitionError error) == null)
+			ServiceDefinitionError error;
+			if (ServiceTypeInfo.TryParse(name, x => s_validNameRegex.IsMatch(x) ? new ServiceDtoInfo(x) : null, position, out error) == null)
```
Comment (ejball) [inline src/Facility.Definition/ServiceDefinitionUtility.cs]: I am eager to upgrade the compiler, build environment, publishing strategy, etc. to something like https://github.com/Faithlife/Parsing but haven't yet found the time.

File: SolutionInfo.cs [hunk 11]
```diff
@@ -1,6 +1,6 @@
 ﻿using System.Reflection;
 
-[assembly: AssemblyVersion("1.3.0.0")]
+[assembly: AssemblyVersion("2.0.0.0")]
```
Comment (ejball) [inline SolutionInfo.cs]: I'd like to avoid a major change for now if possible.

File: src/Facility.Definition/Fsd/FsdParser.cs [hunk 12]
```diff
@@ -68,11 +85,20 @@ public ServiceInfo ParseDefinition(NamedText source)
 			}
 
 			source = new NamedText(source.Name, string.Join("\n", definitionLines));
+			service = null;
 
-			ServiceInfo service;
 			try
 			{
-				service = FsdParsers.ParseDefinition(source, remarksSections);
+				service = FsdParsers.ParseDefinition(source, remarksSections, validationMode: ValidationMode.Return);
```
Comment (ejball) [inline src/Facility.Definition/Fsd/FsdParser.cs]: Not a big deal, but the named parameter is less useful now that the validation mode is strongly typed.

File: src/Facility.Definition/ServiceAttributeInfo.cs [hunk 13]
```diff
@@ -20,8 +21,13 @@ public ServiceAttributeInfo(string name, IEnumerable<ServiceAttributeParameterIn
 			Parameters = parameters.ToReadOnlyList();
 			Position = position;
 
-			ServiceDefinitionUtility.ValidateName(Name, Position);
-			ServiceDefinitionUtility.ValidateNoDuplicateNames(Parameters, "attribute parameter");
+			this.Validate(validationMode);
+		}
+
+		IEnumerable<ServiceDefinitionError> IValidatable.Validate()
```
Comment (ejball) [inline src/Facility.Definition/ServiceAttributeInfo.cs:35]: Why implicit?

File: SolutionInfo.cs (hunk 11 above)
Comment (ddunkin) [reply SolutionInfo.cs]: New constructor parameters and making `position` required for `ServiceDefinitionException`.

File: src/Facility.Definition/Fsd/FsdParser.cs (hunk 12 above)
Comment (ddunkin) [reply src/Facility.Definition/Fsd/FsdParser.cs]: True.

File: src/Facility.Definition/ServiceAttributeInfo.cs (hunk 13 above)
Comment (ddunkin) [reply src/Facility.Definition/ServiceAttributeInfo.cs:35]: `IValidatable` isn't public, so it's hiding the validation mechanics from the public API. It could be part of the API if that's useful.

File: SolutionInfo.cs (hunk 11 above)
Comment (ejball) [reply SolutionInfo.cs]: So, if we introduce constructor overloads (instead of the default parameter) and allow position to be null (even if we never use it that way) we can put off the major version change? I'd like to avoid bumping it until there's a breaking change to `Facility.Core` if at all possible (not to mention generally speaking).

File: src/Facility.Definition/ServiceAttributeInfo.cs (hunk 13 above)
Comment (ejball) [reply src/Facility.Definition/ServiceAttributeInfo.cs:35]: Ah, cool, I forgot to check whether `IValidatable` was public; I'm fine with it as-is.

**Review by ejball — APPROVED** [review]
